In [312]:
# Load the packages
import pandas as pd
import altair as alt

### 1. Load the Data

In this step, we load the datasets into Python using the `pandas` library to demonstrate that the dataset can be loaded into Python.
We read the `players.csv` and `sessions.csv` files, which contain essential details about player characteristics and session activities, respectively.

The code used for this step accomplishes the following:
- It imports the data from the relative file paths into pandas DataFrames.
- It displays the first three rows of each dataset to provide a quick overview of the data structure and contents, allowing us to verify that the datasets have been loaded correctly.

In [313]:
# 1. Load the datasets
players = pd.read_csv('data\players.csv')
sessions = pd.read_csv('data\sessions.csv')

### 2. Wrangle and Clean the Data to the Format Necessary for the Planned Analysis

#### Data Wrangling Steps:
1. Convert All Data into Proper Data Type

2. Verify the Integrity of the Data

3. Handling Missing Data and Dropping Unnecessary Columns 

4. Combine two dataframe

#### 2-1. Convert All Data into Proper Data Type: 

Proper data types are essential for efficient data processing and accurate analysis. This step ensures that each column in our datasets is stored in the most appropriate format, reflecting the nature of the data and optimizing for both memory usage and processing speed. We will convert date columns to datetime objects and other relevant columns to categorical or numerical types based on their content and role in our analysis.


In [314]:
## check the origin data type
print(players.dtypes)
print()
print(sessions.dtypes)

experience           object
subscribe              bool
hashedEmail          object
played_hours        float64
name                 object
gender               object
age                   int64
individualId        float64
organizationName    float64
dtype: object

hashedEmail             object
start_time              object
end_time                object
original_start_time    float64
original_end_time      float64
dtype: object


In [315]:
# Converting data types in 'players'
players['experience'] = players['experience'].astype('category')
players['gender'] = players['gender'].astype('category')

# Converting data types in 'sessions'
sessions['start_time'] = pd.to_datetime(sessions['start_time'],dayfirst=True)
sessions['end_time'] = pd.to_datetime(sessions['end_time'],dayfirst=True)

## check the origin data type again
print(players.dtypes)
print()
print(sessions.dtypes)

experience          category
subscribe               bool
hashedEmail           object
played_hours         float64
name                  object
gender              category
age                    int64
individualId         float64
organizationName     float64
dtype: object

hashedEmail                    object
start_time             datetime64[ns]
end_time               datetime64[ns]
original_start_time           float64
original_end_time             float64
dtype: object


By converting `experience` and `gender` in the `players` dataset to categorical types, we enhance the efficiency of our data storage and simplify the analysis involving these variables. 

By converting `start_time` and `end_time` from the `sessions` dataset to datetime, we enhance accurate and efficient time-based calculations. This adjustment ensures that our data handling is robust and that our analyses will be based on correctly formatted data, enabling precise and reliable results.


#### 2-2. Verifying the Integrity of the Data

Ensuring data integrity is a critical step before having any other data analysis. We need to check for missing values across the datasets. Missing data may significantly impact the process of the analysis, leading to biased or incorrect conclusions if not properly addressed. By identifying missing values before further analysis, we can decide on appropriate strategies for handling them, such as imputation or removal, ensuring a robust dataset for subsequent analyses.

In [316]:
# 2. verifying the integrity of the data

## Check for missing values in each column
missing_data_counts_players = players.isnull().sum()
missing_data_counts_sessions = sessions.isnull().sum()
print(missing_data_counts_players)
print()
print(missing_data_counts_sessions)

experience            0
subscribe             0
hashedEmail           0
played_hours          0
name                  0
gender                0
age                   0
individualId        196
organizationName    196
dtype: int64

hashedEmail            0
start_time             0
end_time               2
original_start_time    0
original_end_time      2
dtype: int64


By checking for missing values in each column of the `players` dataset, we can see that most columns in the `players` dataset are complete except for `individualId` and `organizationName`, which are entirely missing. This indicates that these columns may not provide any useful information for our analysis, as they contain no data at all. So we will drop them in the following wrangling.

By checking for missing values in each column of the `players` dataset, we can see that the `end_time` and `original_end_time` columns each have 2 missing entries, suggesting minor issues with data recording for these specific sessions. Considering the number of missing data is small, we may fill missing `end_time` with the start_time plus the average session duration. The `original_end_time` is not needed for our future analysis and will be drop in following steps, so we can ignore it.


#### 2-3. Handling Missing Data and Dropping Unnecessary Columns

In the `players` DataFrame, we can identify two completely empty columns: `individualId` and `organizationName`. Since these columns contain no useful data, we will drop them from the DataFrame.

In the `sessions` DataFrame, there is missing values in the `end_time` column. Considering the number of observation that has missing data is quite small, we can simply drop them.
Also, we will drop `original_start_time` and `original_end_time` from the `sessions` DataFrame since these columns are not necessary for our analysis.

In [317]:
# handle missing data, drop needless columns

## drop players 's empty column ['individualId', 'organizationName']
players.drop(columns=['individualId', 'organizationName'],inplace = True)

# # Calculate the average duration in seconds (if time is in datetime format)
# average_duration = (sessions['end_time'] - sessions['start_time']).mean()

'''We have two choice here, so we need to pick one'''
# # Iterate over the DataFrame rows using iterrows()
# for index, row in sessions.iterrows():
#     # Check if 'end_time' is NaN
#     if pd.isnull(row['end_time']):
#         # Impute missing 'end_time' by adding the average duration to 'start_time'
#         sessions.at[index, 'end_time'] = row['start_time'] + average_duration

## drop sessions 's empty column ['individualId', 'organizationName']
sessions.drop(columns=['original_start_time', 'original_end_time'],inplace = True)

# drop the observation that contains missing valuable
sessions = sessions.dropna(subset=['end_time'])

## Check for missing values in each column again
missing_data_counts_players = players.isnull().sum()
missing_data_counts_sessions = sessions.isnull().sum()
print(missing_data_counts_players)
print()
print(missing_data_counts_sessions)

experience      0
subscribe       0
hashedEmail     0
played_hours    0
name            0
gender          0
age             0
dtype: int64

hashedEmail    0
start_time     0
end_time       0
dtype: int64


#### 2-4. Combining DataFrames and Analyzing Session Durations

In [318]:
# Calculate session durations in hours
sessions['session_duration'] = (sessions['end_time'] - sessions['start_time']).dt.total_seconds() / 3600

# Aggregate total session time per player
total_session_time = sessions.groupby('hashedEmail')['session_duration'].sum()

# Merge this with player data
players = players.merge(total_session_time, how='left', left_on='hashedEmail', right_index=True)

# Compare the calculated total session time with 'played_hours'
players['discrepancy'] = players['played_hours'] - players['session_duration']

# Fill NaN only in specific columns to avoid issues with categorical data
players[['session_duration', 'discrepancy']] = players[['session_duration', 'discrepancy']].fillna(0)
players

# Check discrepancies
players

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  session_duration  \
0            30.3     Morgan               Male    9         33.650000   
1             3.8  Christian               Male   17          4.250000   
2             0.0      Blake               Male   17          0.083333   
3             0.7      Flora             Female   21          0.833333   
4             0.1      Kylie               Male   21          0.150000   
..            ...        ...                ...  ...               ...   
191           0.0     Bailey             Female   17          0.000000   
192           0.3     Pascal               Male   22          0.350000   
193           0.0      Dylan  Prefer not to say   17          0.083333   
194           2.3     Harlow               Male   17          2.983333   
195           0.2      Ahmed              Other   91          0.250000   

     discrepancy  
0      -3.350000  
1      -0.450000  
2      -0.083333  
3      -0.133333  
4      -0.050000  
..           ...  
191     0.000000  
192    -0.050000  
193    -0.083333  
194    -0.683333  
195    -0.050000  

[196 rows x 9 columns]

### There may be big changes in the following code, so Remember to come back and have a look. All following codes will be general and should be robust enough to handle the possible change in dataframe players
# !!!
### btw, i really don't like matching codes with the requirements, so i just go and write all the codes, let's look back in the future to decide whether they are needed and which part should they belong.

In [319]:
# drop hashedEmail, name
players.drop(columns=['hashedEmail',"name","discrepancy"],inplace = True)
players

experience  subscribe  played_hours             gender  age  \
0          Pro       True          30.3               Male    9   
1      Veteran       True           3.8               Male   17   
2      Veteran      False           0.0               Male   17   
3      Amateur       True           0.7             Female   21   
4      Regular       True           0.1               Male   21   
..         ...        ...           ...                ...  ...   
191    Amateur       True           0.0             Female   17   
192    Veteran      False           0.3               Male   22   
193    Amateur      False           0.0  Prefer not to say   17   
194    Amateur      False           2.3               Male   17   
195        Pro       True           0.2              Other   91   

     session_duration  
0           33.650000  
1            4.250000  
2            0.083333  
3            0.833333  
4            0.150000  
..                ...  
191          0.000000  
192          0.350000  
193          0.083333  
194          2.983333  
195          0.250000  

[196 rows x 6 columns]

In [ ]:
players_copy = players.loc[:]

experience  subscribe  played_hours             gender  age  \
0          Pro       True          30.3               Male    9   
1      Veteran       True           3.8               Male   17   
2      Veteran      False           0.0               Male   17   
3      Amateur       True           0.7             Female   21   
4      Regular       True           0.1               Male   21   
..         ...        ...           ...                ...  ...   
191    Amateur       True           0.0             Female   17   
192    Veteran      False           0.3               Male   22   
193    Amateur      False           0.0  Prefer not to say   17   
194    Amateur      False           2.3               Male   17   
195        Pro       True           0.2              Other   91   

     session_duration  
0           33.650000  
1            4.250000  
2            0.083333  
3            0.833333  
4            0.150000  
..                ...  
191          0.000000  
192          0.350000  
193          0.083333  
194          2.983333  
195          0.250000  

[196 rows x 6 columns]

### Use one-hot encoding to turn categorical variables into vectors

In [321]:
# Applying one-hot encoding to 'experience' and 'gender'
experience_dummies = pd.get_dummies(players['experience'], prefix='experience').astype(int)
gender_dummies = pd.get_dummies(players['gender'], prefix='gender').astype(int)

print(experience_dummies)
print(gender_dummies)

     experience_Amateur  experience_Beginner  experience_Pro  \
0                     0                    0               1   
1                     0                    0               0   
2                     0                    0               0   
3                     1                    0               0   
4                     0                    0               0   
..                  ...                  ...             ...   
191                   1                    0               0   
192                   0                    0               0   
193                   1                    0               0   
194                   1                    0               0   
195                   0                    0               1   

     experience_Regular  experience_Veteran  
0                     0                   0  
1                     0                   1  
2                     0                   1  
3                     0                   0  
4

In [322]:
# Since 'subscribe' only contains two types (True and False), we can convert it directly to an integer.
players['subscribe'] = players['subscribe'].astype(int)  # Replaces True with 1 and False with 0

# Concatenate the new columns back to the original dataframe
players = pd.concat([players, experience_dummies, gender_dummies], axis=1)

# Drop the original 'experience' and 'gender' columns as they are now encoded
players.drop(['experience', 'gender'], axis=1, inplace=True)

players

subscribe  played_hours  age  session_duration  experience_Amateur  \
0            1          30.3    9         33.650000                   0   
1            1           3.8   17          4.250000                   0   
2            0           0.0   17          0.083333                   0   
3            1           0.7   21          0.833333                   1   
4            1           0.1   21          0.150000                   0   
..         ...           ...  ...               ...                 ...   
191          1           0.0   17          0.000000                   1   
192          0           0.3   22          0.350000                   0   
193          0           0.0   17          0.083333                   1   
194          0           2.3   17          2.983333                   1   
195          1           0.2   91          0.250000                   0   

     experience_Beginner  experience_Pro  experience_Regular  \
0                      0               1                   0   
1                      0               0                   0   
2                      0               0                   0   
3                      0               0                   0   
4                      0               0                   1   
..                   ...             ...                 ...   
191                    0               0                   0   
192                    0               0                   0   
193                    0               0                   0   
194                    0               0                   0   
195                    0               1                   0   

     experience_Veteran  gender_Agender  gender_Female  gender_Male  \
0                     0               0              0            1   
1                     1               0              0            1   
2                     1               0              0            1   
3                     0               0              1            0   
4                     0               0              0            1   
..                  ...             ...            ...          ...   
191                   0               0              1            0   
192                   1               0              0            1   
193                   0               0              0            0   
194                   0               0              0            1   
195                   0               0              0            0   

     gender_Non-binary  gender_Other  gender_Prefer not to say  \
0                    0             0                         0   
1                    0             0                         0   
2                    0             0                         0   
3                    0             0                         0   
4                    0             0                         0   
..                 ...           ...                       ...   
191                  0             0                         0   
192                  0             0                         0   
193                  0             0                         1   
194                  0             0                         0   
195                  0             1                         0   

     gender_Two-Spirited  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
191                    0  
192                    0  
193                    0  
194                    0  
195                    0  

[196 rows x 16 columns]

In [323]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.model_selection import train_test_split
from sklearn.compose import make_column_selector
from sklearn import set_config
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.metrics import make_scorer, root_mean_squared_error
import numpy as np

In [324]:
# Output dataframes instead of arrays
set_config(transform_output="pandas")

# Set the random seed for reproducibility
np.random.seed(42) # 42 is the answer to the “ultimate question of life, the universe, and everything,”(Just Joking)

In [325]:
# Spliting players into train set and test set
players_train, players_test = train_test_split(
    players, train_size=0.7
)
players_train.drop(['experience_Pro', 'gender_Other'], axis=1, inplace=True)
players_test.drop(['experience_Pro', 'gender_Other'], axis=1, inplace=True)

y_train = players_train['played_hours']
X_train = players_train.drop(['played_hours'], axis=1)

y_test = players_test['played_hours']
X_test = players_test.drop(['played_hours'], axis=1)

In [326]:
# # Example DataFrame setup
# # players = pd.read_csv("your_data.csv")

# # Define categories and their dummies
# categorical_groups = {
#     'gender': [col for col in players.columns if col.startswith('gender')],
#     'experience': [col for col in players.columns if col.startswith('experience')],
#     'gender': [col for col in players.columns if col.startswith('subscribe')],
# }

# # Start with initial numerical columns that are not part of any categorical dummy groups
# selected = [col for col in players.columns if col not in sum(categorical_groups.values(), []) and col != 'played_hours']

# # Define a function to calculate RMSE
# # def rmse_scorer(model, X, y):
# #     y_pred = model.predict(X)
# #     return np.sqrt(mean_squared_error(y, y_pred))
# def rmse(y_true, y_pred):
#     return root_mean_squared_error(y_true, y_pred)

# # Feature selection process
# for category, features in categorical_groups.items():
#     # Temporarily add this category's features to the selected list
#     trial_features = selected + features
#     X = players[trial_features]
#     y = players['played_hours']
    
#     # Model and scoring
#     lm_preprocessor = make_pipeline(StandardScaler(), LinearRegression())
#     score = np.mean(cross_val_score(lm_preprocessor, X, y, cv=10, scoring=make_scorer(rmse_scorer, greater_is_better=False)))
    
#     # Decide whether to permanently add this group of features
#     print(f"Testing inclusion of {category}: RMSE = {score}")
#     include_decision = input(f"Include {category} features? (yes/no): ")
#     if include_decision.lower() == 'yes':
#         selected.extend(features)  # Only add permanently if confirmed

# print("Selected features:", selected)

# # names = [col for col in players.columns if col != 'played_hours']

# # accuracy_dict = {"size": [], "selected_predictors": [], "accuracy": []}

# # # Store the total number of predictors
# # n_total = len(names)

# # # Start with an empty list of selected predictors
# # selected = []

# # # Define a custom scorer function
# # def rmse(y_true, y_pred):
# #     return root_mean_squared_error(y_true, y_pred)

# # # Create the scorer object
# # rmse_scorer = make_scorer(rmse, greater_is_better=False)

# # # Example using LinearRegression
# # lm = LinearRegression()

# # # Create the pipeline
# # lm_preprocessor = make_pipeline(StandardScaler(), lm)

# # # For every possible number of predictors
# # for i in range(1, n_total + 1):
# #     accs = np.zeros(len(names))
# #     # For every possible predictor to add
# #     for j in range(len(names)):
# #         # Add remaining predictor j to the model
# #         X = players[selected + [names[j]]]
# #         y = players['played_hours']

# #         # Calculate cross-validated R-squared score for this set of predictors
# #         scores = cross_val_score(lm_preprocessor, X, y, cv=10, scoring=rmse_scorer)
# #         accs[j] = np.mean(scores)

# #     # Get the best new set of predictors that maximize the cross-validated R-squared
# #     best_set = selected + [names[accs.argmax()]]

# #     # Store the results for this round of forward selection
# #     accuracy_dict["size"].append(i)
# #     accuracy_dict["selected_predictors"].append(", ".join(best_set))
# #     accuracy_dict["accuracy"].append(accs.max())

# #     # Update the selected & remove the chosen predictor from the list
# #     selected = best_set
# #     names.remove(names[accs.argmax()])

# # # Create a DataFrame to show the results
# # accuracies = pd.DataFrame(accuracy_dict)
# # print(accuracies)


In [327]:
# names = [col for col in players.columns if col != 'played_hours']

# accuracy_dict = {"size": [], "selected_predictors": [], "accuracy": []}

# # Store the total number of predictors
# n_total = len(names)

# # Start with an empty list of selected predictors
# selected = []

# # Define a custom scorer function
# def rmse(y_true, y_pred):
#     return root_mean_squared_error(y_true, y_pred)

# # Create the scorer object
# rmse_scorer = make_scorer(rmse, greater_is_better=False)

# # Example using LinearRegression
# lm = LinearRegression()

# # Create the pipeline
# lm_preprocessor = make_pipeline(StandardScaler(), lm)

# # For every possible number of predictors
# for i in range(1, n_total + 1):
#     accs = np.zeros(len(names))
#     # For every possible predictor to add
#     for j in range(len(names)):
#         # Add remaining predictor j to the model
#         X = players[selected + [names[j]]]
#         y = players['played_hours']

#         # Calculate cross-validated R-squared score for this set of predictors
#         scores = cross_val_score(lm_preprocessor, X, y, cv=10, scoring=rmse_scorer)
#         accs[j] = np.mean(scores)

#     # Get the best new set of predictors that maximize the cross-validated R-squared
#     best_set = selected + [names[accs.argmax()]]

#     # Store the results for this round of forward selection
#     accuracy_dict["size"].append(i)
#     accuracy_dict["selected_predictors"].append(", ".join(best_set))
#     accuracy_dict["accuracy"].append(accs.max())

#     # Update the selected & remove the chosen predictor from the list
#     selected = best_set
#     names.remove(names[accs.argmax()])

# # Create a DataFrame to show the results
# accuracies = pd.DataFrame(accuracy_dict)
# print(accuracies)

In [328]:
# Initialize the Linear Regression model
lm = LinearRegression()

lm.fit(
   X_train,  # A single-column data frame
   y_train  # A series
)

# Get coefficients and intercept
coef = lm.coef_
intercept = lm.intercept_

# Assuming you have the names of the features
feature_names = [col for col in X_train.columns]

# Constructing the formula
terms = [f"{coef[i]:.3f} * {feature_names[i]}" for i in range(len(feature_names))]
formula = " + ".join(terms)
linear_model_formula = f"y = {intercept:.3f} + {formula}"

print("Linear Model Formula:")
print(linear_model_formula)

Linear Model Formula:
y = 1.028 + -0.287 * subscribe + -0.008 * age + 0.904 * session_duration + -0.485 * experience_Amateur + -0.005 * experience_Beginner + 0.365 * experience_Regular + -0.078 * experience_Veteran + -0.402 * gender_Agender + -0.549 * gender_Female + -0.829 * gender_Male + -0.166 * gender_Non-binary + -0.673 * gender_Prefer not to say + -0.338 * gender_Two-Spirited


### 3. Perform a Summary of the Data Set that is Relevant for Exploratory Data Analysis Related to the Planned Analysis

In [331]:
#  Summarise Statistics for numerical variables
summary_stats = players_copy.describe()
summary_stats

played_hours         age  session_duration
count    196.000000  196.000000        196.000000
mean       5.845918   21.280612          6.629762
std       28.357343    9.706346         31.310015
min        0.000000    8.000000          0.000000
25%        0.000000   17.000000          0.000000
50%        0.100000   19.000000          0.166667
75%        0.600000   22.000000          0.820833
max      223.100000   99.000000        244.516667

The summary statistics provide valuable insights into the distribution of `played_hours` and `age` within the `players` dataset:

**Played Hours:**
- **Mean**: 0.99 hours—indicating most sessions are short.
- **Standard Deviation**: 3.51—highlighting high variability with some players having significantly longer sessions.
- **Range**: 0 to 30.3 hours, with 25% not engaging and 75% playing less than 0.5 hours.

**Age:**
- **Mean**: The average age of players is approximately 21.38 years.
- **Standard Deviation**: The standard deviation(9.89) is almost half of the mean but does not exceed it, suggesting a relatively uniform distribution around the mean.
- **Range**: 8 to 99 years, with 75% of players under 22 years old, suggesting a predominantly young player base. The oldest player is 99 years old, indicating a broad age range. However, this could potentially be a misentry considering.

In [333]:
# Count frequency for Explanatory Variables

# Categorical variables
experience_counts = players_copy.groupby('experience',observed=False).size().reset_index(name='count')
gender_counts = players_copy.groupby('gender',observed=False).size().reset_index(name='count')
subscribe_counts = players_copy.groupby('subscribe',observed=False).size().reset_index(name='count')

# Numerical variavble `age`
# Define bins and labels
bins = [0, 12, 18, 35, 65, 120]
labels = ['Child(0-12)', 'Teen(12-18)', 'Young Adult(18-35)', 'Adult(35-65)', 'Senior(65-120)']
# Create a new categorical variable 'age_group'
players_copy['age_group'] = pd.cut(players_copy['age'], bins = bins, labels = labels, right = False)
age_counts = players_copy.groupby('age_group',observed=False).size().reset_index(name = 'count')

print(experience_counts,"\n")
print(gender_counts,"\n")
print(subscribe_counts,"\n")
print(age_counts)

  experience  count
0    Amateur     63
1   Beginner     35
2        Pro     14
3    Regular     36
4    Veteran     48 

              gender  count
0            Agender      2
1             Female     37
2               Male    124
3         Non-binary     15
4              Other      1
5  Prefer not to say     11
6       Two-Spirited      6 

   subscribe  count
0      False     52
1       True    144 

            age_group  count
0         Child(0-12)      4
1         Teen(12-18)     83
2  Young Adult(18-35)     99
3        Adult(35-65)      8
4      Senior(65-120)      2


### 4. Create a visualization of the dataset that is relevant for exploratory data analysis related to the planned analysis

##### Histograms for continuous variables - Age and Played Hours

In [ ]:
hist_age = alt.Chart(players).mark_bar().encode(
    alt.X('age', bin = alt.Bin(maxbins = 30), title = 'Age(Years)'),
    alt.Y('count()', title = 'Frequency')
).properties(
    title = 'Histogram of Age',
    width = 360,
    height = 200
)

hist_age

alt.Chart(...)

As is depicted in the histogram, the age groups are concentrated primarily between 15 and 30, with the highest frequency in the 15 to 25 year range. This suggests that the player base is teenagers and young adults.

In [ ]:
hist_played_hours = alt.Chart(players).mark_bar().encode(
    alt.X('played_hours', bin = alt.Bin(maxbins=30), title = 'Played Hours'),
    alt.Y('count()', title = 'Frequency')
).properties(
    title = 'Histogram of Played Time',
    width = 360,
    height = 200
)

hist_played_hours

alt.Chart(...)

As is depicted in the plot, a vast majority have logged very few hours. Specifically, the highest frequency occurs at the lowest time bracket, indicating most players spend less than 10 hours total. This distribution highlights a potential issue in player engagement, as the data suggests many players do not return or play long.

##### Bar Charts for categorical variables - Experience, Gender, Subscribe

In [ ]:
bar_experience = alt.Chart(players).mark_bar().encode(
    x = alt.X('experience:N', sort=['Beginner', 'Amateur', 'Regular', 'Veteran', 'Pro'],
     axis = alt.Axis(labelAngle=0)).title("Experience"),  # Set labelAngle to 0 for horizontal labels
    y = 'count()'
).properties(
    title = 'Player Experience Levels Counts',
    width = 240,
    height = 200
)

bar_experience

alt.Chart(...)

As is depicted in the plot, the majority of players are categorized as 'Amateur', followed by 'Regular' and 'Veteran', indicating a player base with a range of experience but leaning towards newer or moderately experienced players.

In [ ]:
bar_gender = alt.Chart(players).mark_bar().encode(
    x = alt.X('gender:N',sort='-y', axis=alt.Axis(labelAngle=0)).title("Gender"),
    y = 'count()'
).properties(
    title = 'Gender Counts',
    width = 480,
    height = 200
)
bar_gender

alt.Chart(...)

As is depicted in the plot, the distribution of gender among players shows a predominant number of male players compared to other gender identities since the number of male players is over double the number of the next largest group, females.

In [ ]:
bar_subscribe = alt.Chart(players).mark_bar().encode(
    x = alt.X('subscribe:N', axis=alt.Axis(labelAngle=0)).title("Subscribe Status"),
    y = 'count()'
).properties(
    title = 'Subscription Status Counts',
    width = 100,
    height = 200
)
bar_subscribe

alt.Chart(...)

As is depicted in the plot, there is a significantly higher number of subscribed players compared to non-subscribers.

### 5. Perform the data analysis

### 6. Create a visualization of the analysis 